hphp/runtime/vm/jit/vasm-lower.cpp

```diff
@@ -226,28 +277,36 @@ void lower(Vunit& unit, vcallarray& inst, Vlabel b, size_t i) {
       args |= rarg(i);
     }
 
-    v << copyargs{unit.makeTuple(srcs), unit.makeTuple(std::move(dsts))};
+    v << copyargs{env.unit.makeTuple(srcs), env.unit.makeTuple(std::move(dsts))};
     v << callarray{inst.target, args};
     v << unwind{{inst.targets[0], inst.targets[1]}};
   });
 }
 
-void lower(Vunit& unit, defvmsp& inst, Vlabel b, size_t i) {
-  unit.blocks[b].code[i] = copy{rvmsp(), inst.d};
+void lower(VLS& env, defvmsp& inst, Vlabel b, size_t i) {
+  env.unit.blocks[b].code[i] = copy{rvmsp(), inst.d};
 }
-void lower(Vunit& unit, syncvmsp& inst, Vlabel b, size_t i) {
-  unit.blocks[b].code[i] = copy{inst.s, rvmsp()};
+void lower(VLS& env, syncvmsp& inst, Vlabel b, size_t i) {
+  env.unit.blocks[b].code[i] = copy{inst.s, rvmsp()};
 }
 
-void lower(Vunit& unit, defvmretdata& inst, Vlabel b, size_t i) {
-  unit.blocks[b].code[i] = copy{rret_data(), inst.data};
+void lower(VLS& env, defvmretdata& inst, Vlabel b, size_t i) {
+  env.unit.blocks[b].code[i] = copy{rret_data(), inst.data};
+}
+void lower(VLS& env, defvmrettype& inst, Vlabel b, size_t i) {
+  env.unit.blocks[b].code[i] = copy{rret_type(), inst.type};
+}
+void lower(VLS& env, syncvmret& inst, Vlabel b, size_t i) {
+  env.unit.blocks[b].code[i] = copy2{inst.data,   inst.type,
+                                     rret_data(), rret_type()};
 }
-void lower(Vunit& unit, defvmrettype& inst, Vlabel b, size_t i) {
-  unit.blocks[b].code[i] = copy{rret_type(), inst.type};
+void lower(VLS& env, vregallow& inst, Vlabel b, size_t i) {
+  env.allow_vreg = true;
+  lower_impl(env.unit, b, i, [&] (Vout& v) {});
```

Comment:
You're doing this to just kill the `vregallow{}` (and `vregrestrict{}`, below), right?  If so, you should just nop it out instead:

```
env.unit.blocks[b].code[i] = nop{};
```

It avoids a (possibly expensive) shift, which in most implementations of `std::vector` won't (I think) shrink the allocation anyway...